### PR TITLE
Add encode method to convert Sections into GPP String

### PR DESCRIPTION
**Pull request by duduchristian**

### CODE REVIEW

File: util/encoding.go
Comment (hhhjort) [inline util/encoding.go]: It would probably by more efficient to append one mass of bytes, rather than looping over an append statement like this. 
`
size := final - bs.Len()
chuck := make([]byte, size, size)
bs.b = append(bs.b, chunk...)
`
The new array created by make should contain all zeros in all the elements of the array. The documentation isn't 100% clear, but that is how it reads to me.

This may not be a big difference when enlarge is used by the write functions below, as it will only be writing one or two bytes at a time. So I am not too worried about the final form of enlarge currently, but may be more of a factor if we make more expansive write functions in the future. In particular `WriteIntRange()` could probably calculate the total size needed fairly quickly, and then ensure the byte slice in the bit stream has sufficient capacity in one step, rather than depending on a sequence of `enlarge()` calls.

File: encode.go
Comment (hhhjort) [inline encode.go]: It may be more efficient to estimate the final size of the bitstream, and then create a `[]byte` with that capacity. That will save a lot of reallocations and copies as the bitstream is expanded with append statements. Most applications are not so memory constrained that allocating a handful of bytes more than needed per GPP string should be an issue.

**Comment by hhhjort** [comment]
Can you put a note in the README that, in practice, only CMPs should be writing GPP strings as per the IAB documentation? We don't want to encourage people to be generating GPP strings in production when they really should not be.

**Comment by hhhjort** [comment]
Just a friendly ping to see if you have read the comments, and intend to keep working on this PR.

Thanks

**Comment by duduchristian** [comment]
Coming back, still working on this.

> Just a friendly ping to see if you have read the comments, and intend to keep working on this PR.
> 
> Thanks

Coming back, still working~

File: encode.go
Comment (hhhjort) [inline encode.go]: The size should be `len(intRange.Range)`, not the total number of sections.

File: encode_test.go
Comment (hhhjort) [inline encode_test.go:23]: Should probably have a test or two where multiple sections are encoded.

File: util/encoding_test.go
Comment (hhhjort) [inline util/encoding_test.go:111]: The way you have your tests set up, you will always be writing on the even boundaries of your bit units. For example, `TestWriteByte2` will always be writing to even bit indexes. `TestWriteByte8` will always be writing on byte boundaries. You need to mix in some arbitrary offsets to exercise the tests to ensure the functions work properly at arbitrary indexes.

Comment (duduchristian) [reply util/encoding_test.go:111]: refer to TestArbitraryWrite

**Comment by hhhjort** [comment]
It may be interesting to write a test where we decode some GPP strings, then re-encode them and compare to see if we got the original string back. It could uncover a subtle error, or potentially that the spec is under specified such that there are multiple ways to encode a set of data into a string.

File: encode.go
Comment (bsardo) [inline encode.go]: Nitpick: when there aren't any parameters in your error message, I believe it is preferable to use `errors.New` from the `errors` package.

File: util/encoding.go
Comment (bsardo) [inline util/encoding.go]: Nitpick: when there aren't any parameters in your error message, I believe it is preferable to use `errors.New` from the `errors` package.

File: encode.go
Comment (bsardo) [inline encode.go]: I suggest we use another method to track this instead of putting TODOs in the code. If this is something that needs to be done in a separate PR, perhaps we can track this in a different way. Maybe create an issue? @hhhjort thoughts?

I noticed the `Encode` methods for some of the regs has code to exit early if this value is false but it doesn't look like there is currently any way to take that code path so tests should probably be added for those lines when this TODO is addressed.

File: encode_test.go
Comment (bsardo) [inline encode_test.go]: Nitpick: can you put `uspca.USPCA{` on the next line and indent so all of the sections line up so it is easier to read? The same applies to other test cases you have here.

File: util/encoding_test.go
Comment (bsardo) [inline util/encoding_test.go]: I understand that there is value here as this test can be used to verify writes off of byte boundaries, however, I think there is a better way to handle this. The exact tests executed are unpredictable given the use of the random generator. I suggest deleting this test and instead specifying the exact off byte boundary test cases you want to cover.
If you follow my approach suggested in another comment that asks for the test cases to be declared in each test rather than declared globally in the file, you can easily add additional test cases to each of the write function tests to handle these off byte boundary scenarios.
It's a bit more verbose but predictable and easier to maintain since each test is isolated.

**Comment by hhhjort** [comment]
Sorry we have been a bit busy and haven't responded in a bit. This PR seems to be in a good state, and we should get the final review in soon.

File: test.sh
Comment (bsardo) [inline test.sh:1]: It appears you want this file to be executable however it it is read-only.

Also, we should probably update the `.gitignore` to include `release/` since you're creating files in that directory when this script is executed.

Comment (bsardo) [inline test.sh:1]: Can you also add `.idea/` and `.vscode/` to `.gitignore`?

File: util/encoding_test.go
Comment (bsardo) [inline util/encoding_test.go:407]: Can you please add a `name` field to each test case describing what the test case covers and make use of `t.Run`:
```
for name, test := range tests {
		t.Run(name, func(t *testing.T) {
```

File: util/encoding_test.go
Comment (bsardo) [inline util/encoding_test.go]: Same comment as above. Can you please add a `name` field to each test case describing what the test case covers and make use of `t.Run`:
```
for name, test := range tests {
		t.Run(name, func(t *testing.T) {
```

File: util/encoding_test.go
Comment (bsardo) [inline util/encoding_test.go]: Nitpick: can you make this condition based on the expectation `c.err` instead of `err`? Actually, why is this condition needed at all? Shouldn't we know what the value should be given the test inputs regardless of whether there is an error?

File: util/encoding.go
Comment (bsardo) [inline util/encoding.go]: Some of the functions here only have the name of the function as a comment which I believe was done temporarily to quiet the lint checker. Can you please expand on these comments? IMO functions like this one could especially use more information to inform the reader of intent.

File: util/encoding_test.go
Comment (bsardo) [inline util/encoding_test.go:78]: Can we maybe rewrite this and the other write byte tests so they are isolated making use of `t.Run` with their own test cases (which can explicitly call out byte boundary scenarios) instead of shared ones and explicitly call the function that is under test instead of making a helper function with conditional logic determine what to call? IMO this would make the tests more straightforward and simplify the helper.

How about something like this:
```
func TestWriteByte1(t *testing.T) {
	testCases := []struct{
		name string
		data []byte
	}{
		{
			name: "single_byte",
			data: []byte("a"),
		},
		{
			name: "bytes_string",
			data: []byte("too young too simple sometimes naive"),
		},
		{
			name: "bytes_unicode",
			data: []byte{0x04, 0xa2, 0x03, 0xb1, 0x00, 0x2b},
		},
	}

	for _, tc := range testCases {
		t.Run(tc.name, func(t *testing.T) {
			bs := NewBitStream(nil)
			p := 0
			for p < len(tc.data)*8 {
				dataToWrite := prepareNBits(tc.data, p, 1)
				bs.WriteByte1(byte(dataToWrite))
				p += 1
			}
			assert.Equal(t, tc.data, bs.b)
		})
	}
}
```

In this case I changed `writeNBits` to `prepareNBits`.

**Comment by bsardo** [comment]
Hi @duduchristian, just checking in to see if you have any time to address my comments. Overall, it looks good. No rush and thanks!

**Comment by duduchristian** [comment]
> 

yep, I get some time next week. Just coming back from my holiday hahah.

**Comment by bsardo** [comment]
Hi @duduchristian, just checking in to see if you have some time to address the outstanding comments. No rush 🙂

**Comment by bsardo** [comment]
Hi @duduchristian. It looks like you made some of the updates but there are a few more comments outstanding. Are you planning to address those as well? Thanks!

**Comment by duduchristian** [comment]
> Hi @duduchristian. It looks like you made some of the updates but there are a few more comments outstanding. Are you planning to address those as well? Thanks!

Sorry I was so busy in the previous weeks. It seems I've addressed all the problems mentioned in comments, can you check the code and if there are any problems, make comments just as what you did before. Thank you.

File: .deepsource.toml
Comment (hhhjort) [inline .deepsource.toml]: Is this file supposed to be included? If so, should the `import_root` not be modified to point to this repository?

File: util/encoding.go
Comment (hhhjort) [inline util/encoding.go]: I am not sure what this is calculating or what expThreshold is supposed to be?

Comment (hhhjort) [reply util/encoding.go]: Would it be better to say `size := cap(bs.b) * 2` rather than using this complex function that is creating an int pointer out of a float pointer? This would seem to depend strongly on just how the CPU architecture represents ints and floats in memory. 

Actually you would need to loop in case you need more than one doubling in size.
```suggestion
		if final <= expThreshold {
		        size := expThreshold
		        for size > final*2 {
		                size = size/2
		        }
		        temp = make([]byte, final, size)
```
This would also fix the bitstream starting at some odd size.

File: util/encoding.go
Comment (hhhjort) [inline util/encoding.go]: Might want to add a parenthesis to make the order of operations clearer for humans.
`((final-expThreshold+step-1)/step)*step`
It took me a bit to parse this correctly in my head.

File: util/encoding.go
Comment (hhhjort) [inline util/encoding.go:26]: This is calculating the final position (or size) from where the bitstream pointer currently is. It would be sketchy to write to the middle of a bit stream, but should probably make this clearer in the comment so someone doesn't assume it can divine where the end of the bit stream is.

File: util/encoding.go
Comment (hhhjort) [inline util/encoding.go]: If you are limiting to `fibLen`, then you can just lookup in the `fibLookup[i]` array rather than calling the `fibonacci()` function. The function just adds code to handle the case where the number if greater than `fibLen`.

File: util/encoding.go
Comment (hhhjort) [inline util/encoding.go:214]: Shouldn't Fibonacci enter the name someplace as the GPP spec also defines a non-Fibonacci int range. The non-Fibonacci range will be needed once TCFv2 support is added.

**Comment by bsardo** [comment]
Hi @duduchristian, thanks for responding to our comments. We recently left a few more for you to address when you get a chance. If you don't mind, could you please merge with master instead of force pushing? It will save us a lot of time reviewing. Thanks!

File: util/encoding.go
Comment (hhhjort) [inline util/encoding.go]: This requires some work.

For example, if final = 70 and newCap = 40, then `final > doubleCap` is false, `newCap < 32` is false, so we have `newCap += newCap / 4` which makes  the new newCap = 50. The `make` will then fail as you try to make an array sized larger than its capacity.

File: util/encoding.go
Comment (hhhjort) [reply util/encoding.go:214]: Would you like to comment on the difference between the Fibonacci Int range and the legacy Int Range that is used in TCF. We do plan to add TCFv2 code to this library once we get some time to do it. It hasn't been too urgent as an independent TCFv2 library exists.

**Review by hhhjort — DISMISSED** [review]
Looks good to me now. Will get a second pair of eyes to review before we merge.

File: util/encoding.go
Comment (hhhjort) [inline util/encoding.go]: These calls to `fibonacci()` can also be replaced with array lookups to optimize thingsa little.

File: util/encoding.go
Comment (hhhjort) [reply util/encoding.go:214]: The current name will be an issue when TCFv2 is brought into the project and the original Int range is implemented. The documentation refers to them as `Range (Fibonacci)` and `Range (Int)`, so the current name could confuse people looking for the `Range (Fibonacci)` implementation. This is why the corresponding read function is called `ReadFibonacciRange()`.

File: encode.go
Comment (guscarreon) [inline encode.go:51]: If `id` equals zero in the first loop of this `for` statement, we'll be referencing `Range` at position `-1`. The following corner case indeed makes `Encode(sections []Section)` panic:
```
 16   type gppEncodeTestData struct {
 17       description string
 18       sections    []Section
 19       expected    string
 20   }
 21  
    + type zeroSection struct{}
    + 
    + func (s zeroSection) Encode(b bool) []byte {
    +     return nil
    + }
    + 
    + func (s zeroSection) GetID() constants.SectionID {
    +     return constants.SectionID(0)
    + }
    + 
    + func (s zeroSection) GetValue() string {
    +     return "string"
    + }
    + 
 22   var testData = []gppEncodeTestData{
    +     {
    +         description: "Corner case of a section whose GetID returns 0",
    +         sections:    []Section{zeroSection{}},
    +     },
 23       {
 24           description: "USPCA GPP string encoding",
 25           expected:    "DBABh4A~BlgWEYCY.QA~BSFgmiU",
 26           sections: []Section{
 27               uspca.USPCA{
 28                   CoreSegment: uspca.USPCACoreSegment{
 29                       Version:                     1,
 30                       SaleOptOutNotice:            2,
 31                       SharingOptOutNotice:         1,
 32                       SensitiveDataLimitUseNotice: 1,
 33 *--251 lines: SaleOptOut:                  2,-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
284       },
285   }
286  
287   func TestEncode(t *testing.T) {
288       for _, test := range testData {
289           result, err := Encode(test.sections)
encode_test.go
```
Can we add a check?

**Comment by bsardo** [comment]
Hi @duduchristian. It looks like we're close here. If you could address the remaining comment when you get a chance, that would be great. Thanks!

**Review by hhhjort — DISMISSED** [review]
LGTM

File: util/encoding.go
Comment (guscarreon) [inline util/encoding.go]: Nitpick: "Most strings"

File: encode_test.go
Comment (guscarreon) [inline encode_test.go]: Nitpick: can we set to it's `constants.SectionUSPVA` value instead?

**Review by guscarreon — COMMENTED** [review]
Looks pretty good to me. I just left two very minor comments

**Review by guscarreon — APPROVED** [review]
LGTM

**Review by hhhjort — APPROVED** [review]

**Pull request merged**
